File: apps/dsa-web/src/api/adminUsers.ts
import apiClient from './index';
import { isAxiosApiError } from './error';
import { toCamelCase } from './utils';

export type UserRole = 'admin' | 'user';

export interface AdminUser {
  id: number;
  username: string;
  role: UserRole;
  isActive: boolean;
  /** ISO 时间字符串；账号缺少创建时间时为 null。 */
  createdAt: string | null;
}

export interface AdminUserListResponse {
  users: AdminUser[];
}

export interface AdminCreateUserPayload {
  username: string;
  password: string;
  role: UserRole;
}

export interface AdminUpdateUserPayload {
  role?: UserRole;
  isActive?: boolean;
}

/** GET /api/v1/auth/status 的多用户扩展字段（旧后端可能不返回这些字段）。 */
export interface AuthIdentityStatus {
  authEnabled: boolean;
  loggedIn: boolean;
  multiUser?: boolean;
  username?: string;
  role?: UserRole;
}

/**
 * 提取后端返回的业务错误码（如 last_admin / quota_exceeded / username_taken），
 * 供页面按错误码映射文案。
 */
export function getAdminErrorCode(error: unknown): string | null {
  if (!isAxiosApiError(error)) {
    return null;
  }
  const data = (error as { response?: { data?: unknown } }).response?.data;
  if (data && typeof data === 'object' && typeof (data as { error?: unknown }).error === 'string') {
    return (data as { error: string }).error;
  }
  return null;
}

function toUpdatePayload(payload: AdminUpdateUserPayload): Record<string, unknown> {
  const body: Record<string, unknown> = {};
  if (payload.role !== undefined) body.role = payload.role;
  if (payload.isActive !== undefined) body.isActive = payload.isActive;
  return body;
}

export const adminUsersApi = {
  async listUsers(): Promise<AdminUserListResponse> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/users');
    return toCamelCase<AdminUserListResponse>(response.data);
  },

  async createUser(payload: AdminCreateUserPayload): Promise<AdminUser> {
    const response = await apiClient.post<Record<string, unknown>>('/api/v1/admin/users', payload);
    const data = toCamelCase<{ user: AdminUser | null }>(response.data);
    if (!data.user) {
      throw new Error('create user returned no user payload');
    }
    return data.user;
  },

  /** 返回更新后的用户；目标账号已被删除时后端可能返回 null。 */
  async updateUser(userId: number, payload: AdminUpdateUserPayload): Promise<AdminUser | null> {
    const response = await apiClient.patch<Record<string, unknown>>(
      `/api/v1/admin/users/${userId}`,
      toUpdatePayload(payload),
    );
    const data = toCamelCase<{ user: AdminUser | null }>(response.data);
    return data.user;
  },

  async resetUserPassword(userId: number, newPassword: string): Promise<void> {
    await apiClient.post(`/api/v1/admin/users/${userId}/reset-password`, { newPassword });
  },
};

/**
 * 读取当前登录身份（username / role / multiUser）。
 * 会话状态本身仍以 AuthContext 为准；这里只补充多用户身份字段，
 * 因此放在本模块而不是复用 authApi 的窄类型。
 */
export async function fetchAuthIdentity(): Promise<AuthIdentityStatus> {
  const response = await apiClient.get<Record<string, unknown>>('/api/v1/auth/status');
  return toCamelCase<AuthIdentityStatus>(response.data);
}

File: api/v1/endpoints/admin_users.py
# -*- coding: utf-8 -*-
"""用户管理端点（多用户模式，仅管理员可用）。"""

from __future__ import annotations

import logging
from typing import Optional

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse, Response
from pydantic import BaseModel, Field

from api.deps import require_admin
from src.auth import (
    change_user_account_password,
    get_user_account,
    is_multi_user_enabled,
    list_user_accounts,
    set_user_account_active,
    set_user_account_role,
)

logger = logging.getLogger(__name__)

router = APIRouter(dependencies=[Depends(require_admin)])


class CreateUserRequest(BaseModel):
    """管理员创建账号请求。"""

    username: str = Field(default="", description="Username")
    password: str = Field(default="", description="Initial password")
    role: str = Field(default="user", description="'admin' | 'user'")


class UpdateUserRequest(BaseModel):
    """角色 / 启用状态变更请求。"""

    model_config = {"populate_by_name": True}

    role: Optional[str] = Field(default=None)
    is_active: Optional[bool] = Field(default=None, alias="isActive")


class ResetPasswordRequest(BaseModel):
    """管理员重置密码请求。"""

    new_password: str = Field(default="", alias="newPassword")


def _multi_user_guard() -> Optional[JSONResponse]:
    """多用户模式总闸：关闭时用户管理不可用。"""
    if not is_multi_user_enabled():
        return JSONResponse(
            status_code=403,
            content={"error": "multi_user_disabled", "message": "多用户模式未开启"},
        )
    return None


def _user_payload(user: dict) -> dict:
    return {
        "id": user["id"],
        "username": user["username"],
        "role": user["role"],
        "isActive": user["is_active"],
        "createdAt": user["created_at"].isoformat() + "Z" if user.get("created_at") else None,
    }


@router.get(
    "",
    summary="List users",
    description="List all user accounts (no credential material).",
)
async def admin_list_users():
    guard = _multi_user_guard()
    if guard:
        return guard
    return {"users": [_user_payload(u) for u in list_user_accounts()]}


@router.post(
    "",
    summary="Create user",
    description="Create an account. First account auto-becomes admin when no admin exists.",
)
async def admin_create_user(body: CreateUserRequest):
    guard = _multi_user_guard()
    if guard:
        return guard

    from src.auth import create_user_account

    error_code, user = create_user_account(
        (body.username or "").strip(),
        body.password or "",
        role=(body.role or "user").strip() or "user",
    )
    if error_code == "invalid_username":
        return JSONResponse(
            status_code=400,
            content={"error": "invalid_username", "message": "用户名需为 3-32 位字母、数字、下划线或连字符"},
        )
    if error_code == "invalid_password":
        return JSONResponse(
            status_code=400,
            content={"error": "invalid_password", "message": "密码至少 6 位"},
        )
    if error_code == "username_taken":
        return JSONResponse(
            status_code=409,
            content={"error": "username_taken", "message": "用户名已被使用"},
        )
    if error_code == "max_users":
        return JSONResponse(
            status_code=403,
            content={"error": "quota_exceeded", "message": "已达到用户数量上限（MULTI_USER_MAX_USERS）"},
        )
    if error_code or user is None:
        return JSONResponse(
            status_code=500,
            content={"error": "internal_error", "message": "Failed to create account"},
        )
    return JSONResponse(status_code=201, content={"user": _user_payload(user)})


@router.patch(
    "/{user_id}",
    summary="Update user role/active",
    description="Change role or enable/disable. Bumps token_version; protects the last active admin.",
)
async def admin_update_user(user_id: int, body: UpdateUserRequest):
    guard = _multi_user_guard()
    if guard:
        return guard

    if body.role is None and body.is_active is None:
        return JSONResponse(
            status_code=400,
            content={"error": "nothing_to_update", "message": "未提供要更新的字段"},
        )

    if body.role is not None:
        error_code, user = set_user_account_role(user_id, (body.role or "").strip())
        if error_code:
            return _update_error_response(error_code)

    if body.is_active is not None:
        error_code, user = set_user_account_active(user_id, bool(body.is_active))
        if error_code:
            return _update_error_response(error_code)

    refreshed = get_user_account(user_id)
    return {"user": _user_payload(refreshed) if refreshed else None}


def _update_error_response(error_code: str) -> JSONResponse:
    mapping = {
        "user_not_found": (404, "用户不存在"),
        "last_admin": (409, "不能降级或禁用最后一个可用管理员"),
        "invalid_role": (400, "角色仅支持 admin / user"),
        "storage_error": (500, "更新失败"),
    }
    status, message = mapping.get(error_code, (500, "更新失败"))
    return JSONResponse(status_code=status, content={"error": error_code, "message": message})


@router.post(
    "/{user_id}/reset-password",
    summary="Reset user password",
    description="Admin resets a user's password. Invalidates that user's existing sessions.",
)
async def admin_reset_password(user_id: int, body: ResetPasswordRequest):
    guard = _multi_user_guard()
    if guard:
        return guard

    err = change_user_account_password(user_id, body.new_password or "")
    if err and err.startswith("密码至少"):
        return JSONResponse(
            status_code=400,
            content={"error": "invalid_password", "message": err},
        )
    if err == "用户不存在":
        return JSONResponse(
            status_code=404,
            content={"error": "user_not_found", "message": err},
        )
    if err:
        return JSONResponse(
            status_code=500,
            content={"error": "internal_error", "message": err},
        )
    return Response(status_code=204)
